README.md
# signal-sidecar
sidecar that reports detailed health information from a jitsi signal node

## endpoints

* `/health` responds with 200 if this sidecar is reachable
* `/signal/report` responds with 200 if the signal node is healthy and 500 if not

## configuration

* `HTTP_PORT`: port for REST calls [6000]
* `HA_PORT`: TCP port for HAProxy TCP mode [6060] `[TO BE IMPLEMENTED]`
* `LOG_LEVEL`: debug, info, warn, or error [info]
* `JICOFO_ORIG`: origin for jicofo [http://localhost:8888]
* `PROSODY_ORIG`: origin for prosody [http://localhost:5280]
* `STATUS_PATH`: file for ready/drain status [/etc/jitsi/shard-status]
* `POLLING_INTERVAL`: number of seconds between polls [5]
* `PARTIPANT_MAX`: max number of participants before node should be drained [5000]

## /signal/report response json

```
{
    "healthy": [boolean],                              // overall signal node health
    "status": [ready|drain|unknown],                   // drain state of node
    "services": {
        "jicofoReachable": [boolean]                   // jicofo health http reachable
        "jicofoStatusCode": [http status or 0],        // http code from jicofo
        "jicofoStatsReachable": [boolean]              // jicofo health http reachable
        "jicofoStatusStatusCode": [http status or 0],  // http code from jicofo
        "prosodyReachable": [boolean],                 // prosody health http reachable
        "prosodyStatusCode": [http status code or 0],  // http code from prosody
        "statusFileFound": [boolean],                  // was the status file found
        "statusFileContents": [stuff],                 // contents of the status file
    },
    "stats": {
        "jicofoParticipants": [number]                 // number of jicofo participants
        "jicofoConferences": [number]                  // number of jicofo conferences
    }
}
```
## debian build command

> dpkg-buildpackage -A -rfakeroot -us -uc


debian/changelog
signal-sidecar (0.0.1-1) unstable; urgency=medium

  * Initial release. Build from git: 06de991

 -- Jitsi Team <dev@jitsi.org>  Fri, 21 May 2021 21:42:33 +0000


src/app.ts
import config from './config';
import express from 'express';
import logger from './logger';
import HealthCollector, { HealthReport, HealthCollectorOptions } from './health_collector';

logger.info('signal-sidecar startup', { config });

// health polling
const healthCollectorOptions: HealthCollectorOptions = {
    jicofoHealthUrl: config.JicofoOrig + '/about/health',
    jicofoStatsUrl: config.JicofoOrig + '/stats',
    prosodyHealthUrl: config.ProsodyOrig + '/http-bind',
    statusFilePath: config.StatusPath,
    healthPollingInterval: config.PollingInterval,
};

const healthCollector = new HealthCollector(healthCollectorOptions);
const initHealthReport = <HealthReport>{
    healthy: false,
    status: 'unknown',
    services: {
        jicofoReachable: false,
        jicofoStatusCode: 0,
        jicofoStatsReachable: false,
        jicofoStatsStatusCode: 0,
        prosodyReachable: false,
        prosodyStatusCode: 0,
        statusFileFound: false,
        statusFileContents: '',
    },
    stats: {
        jicofoParticipants: null,
        jicofoConferences: null,
    },
};
let healthReport: HealthReport = initHealthReport;

async function pollForHealth() {
    logger.debug('entering pollForHealth', { report: healthReport });
    try {
        healthReport = await healthCollector.updateHealthReport();
    } catch (err) {
        logger.error('pollForHealth error', { err });
        healthReport = initHealthReport;
    }
    setTimeout(pollForHealth, healthCollectorOptions.healthPollingInterval * 1000);
}
pollForHealth();

// express handlers
const app = express();

async function healthReportHandler(req: express.Request, res: express.Response) {
    if (healthReport) {
        res.status(200);
        if (!healthReport.healthy) {
            res.status(500);
        }
        res.send(JSON.stringify(healthReport));
    } else {
        res.sendStatus(500);
    }
}

async function signalHealthHandler(req: express.Request, res: express.Response) {
    if (healthReport) {
        res.status(200);
        if (!healthReport.healthy) {
            res.status(503);
            res.send('NOT_OK');
        } else {
            res.send('OK');
        }
    } else {
        res.status(500);
        res.send('NOT_OK');
    }
}

// health of the signal-sidecar itself
app.get('/health', (req: express.Request, res: express.Response) => {
    res.sendStatus(200);
});

// overall health of signal node, intended to be a public endpoint
app.get(['/about/health', '/signal/health'], async (req, res, next) => {
    try {
        await signalHealthHandler(req, res);
    } catch (err) {
        next(err);
    }
    res.sendStatus(200);
});

// detailed health report intended for internal use for load balancing
app.get('/signal/report', async (req, res, next) => {
    try {
        await healthReportHandler(req, res);
    } catch (err) {
        next(err);
    }
});

app.listen(config.HTTPServerPort, () => {
    logger.info(`signal-sidecar listening on :${config.HTTPServerPort}`);
});


src/health_collector.ts
import config from './config';
import logger from './logger';
import got from 'got';
import { readFileSync } from 'fs';

export interface HealthData {
    reachable: boolean;
    code: number;
    contents: string;
}

export interface HealthReport {
    healthy: boolean;
    status: string;
    services: {
        jicofoReachable: boolean;
        jicofoStatusCode: number;
        jicofoStatsReachable: boolean;
        jicofoStatsStatusCode: number;
        prosodyReachable: boolean;
        prosodyStatusCode: number;
        statusFileFound: boolean;
        statusFileContents: string;
    };
    stats: {
        jicofoParticipants: number;
        jicofoConferences: number;
    };
}

export interface HealthCollectorOptions {
    jicofoHealthUrl: string;
    jicofoStatsUrl: string;
    prosodyHealthUrl: string;
    statusFilePath: string;
    healthPollingInterval: number;
}

export default class HealthCollector {
    private jicofoHealthUrl: string;
    private jicofoStatsUrl: string;
    private prosodyHealthUrl: string;
    private statusFilePath: string;
    private requestTimeout: number;
    private requestRetryCount: number;

    constructor(options: HealthCollectorOptions) {
        this.jicofoHealthUrl = options.jicofoHealthUrl;
        this.jicofoStatsUrl = options.jicofoStatsUrl;
        this.prosodyHealthUrl = options.prosodyHealthUrl;
        this.statusFilePath = options.statusFilePath;
        this.requestTimeout = 3 * 1000;
        this.requestRetryCount = 2;

        this.updateHealthReport = this.updateHealthReport.bind(this);
    }

    async checkHealthHttp(url: string, method = 'GET'): Promise<HealthData> {
        logger.debug('health check of ' + url);
        try {
            let got_method = got.get;
            if (method == 'POST') {
                got_method = got.post;
            }
            const response = await got_method(url, {
                responseType: 'text',
                timeout: this.requestTimeout,
                retry: this.requestRetryCount,
                throwHttpErrors: false,
            });
            return <HealthData>{
                reachable: true,
                code: response.statusCode,
                contents: response.body,
            };
        } catch (err) {
            logger.debug('checkHealthHttp failed ERROR ERROR', { err, url });
            return <HealthData>{
                reachable: false,
                code: 0,
                contents: '',
            };
        }
    }

    async readStatusFile(filePath: string): Promise<HealthData> {
        logger.debug('status file check of ' + filePath);

        try {
            const data = readFileSync(filePath, 'utf8');
            return <HealthData>{
                reachable: true,
                code: 1,
                contents: data.trim(),
            };
        } catch (err) {
            logger.warn('readStatusFile failed', { err, path: filePath });
            return <HealthData>{
                reachable: false,
                code: 0,
                contents: '',
            };
        }
    }

    // returns [parsable, # participants, # conferences]
    readStatsJSON(jstats: string): [boolean, number, number] {
        try {
            const parsed = JSON.parse(jstats);
            const participants = parsed['participants'];
            const conferences = parsed['conferences'];
            return [true, participants, conferences];
        } catch (err) {
            logger.warn('failed to parse jicofo stats json', { err, json: jstats });
            return [false, 0, 0];
        }
    }

    async updateHealthReport(): Promise<HealthReport> {
        // spawn concurrent calls
        const ccalls: Promise<HealthData>[] = [];
        ccalls.push(this.checkHealthHttp(this.jicofoHealthUrl));
        ccalls.push(this.checkHealthHttp(this.jicofoStatsUrl));
        ccalls.push(this.checkHealthHttp(this.prosodyHealthUrl));
        ccalls.push(this.readStatusFile(this.statusFilePath));

        return Promise.all(ccalls).then((results: HealthData[]) => {
            const jicofoReachable = results[0].reachable;
            const jicofoStatusCode = results[0].code;
            const jicofoStatsReachable = results[1].reachable;
            const jicofoStatsStatusCode = results[1].code;
            const jicofoStatsContents = results[1].contents;
            const prosodyHealthReachable = results[2].reachable;
            const prosodyHealthStatusCode = results[2].code;
            const statusFileReachable = results[3].reachable;
            const statusFileContents = results[3].contents;

            const jStats = this.readStatsJSON(jicofoStatsContents);

            let overallhealth = false;
            if (
                jicofoReachable &&
                jicofoStatusCode == 200 &&
                jicofoStatsReachable &&
                jicofoStatsStatusCode == 200 &&
                prosodyHealthReachable &&
                prosodyHealthStatusCode == 200 &&
                statusFileReachable &&
                jStats[0] // stats file parsed successfully
            ) {
                overallhealth = true;
            }

            let overallstatus = statusFileContents;
            if (jStats[1] > config.ParticipantMax) {
                overallstatus = 'drain';
            }

            const report = <HealthReport>{
                healthy: overallhealth,
                status: overallstatus,
                services: {
                    jicofoReachable: jicofoReachable,
                    jicofoStatusCode: jicofoStatusCode,
                    jicofoStatsReachable: jicofoStatsReachable,
                    jicofoStatsStatusCode: jicofoStatsStatusCode,
                    prosodyReachable: prosodyHealthReachable,
                    prosodyStatusCode: prosodyHealthStatusCode,
                    statusFileFound: statusFileReachable,
                    statusFileContents: statusFileContents,
                },
                stats: {
                    jicofoParticipants: jStats[1],
                    jicofoConferences: jStats[2],
                },
            };
            logger.debug('updateHealthReport return', report);
            return report;
        });
    }
}


src/config.ts
import * as dotenv from 'dotenv';
import envalid from 'envalid';

dotenv.config();

const env = envalid.cleanEnv(process.env, {
    HTTP_PORT: envalid.num({ default: 6000 }),
    TCP_PORT: envalid.num({ default: 6060 }),
    LOG_LEVEL: envalid.str({ default: 'info' }),
    JICOFO_ORIG: envalid.str({ default: 'http://localhost:8888' }),
    PROSODY_ORIG: envalid.str({ default: 'http://localhost:5280' }),
    STATUS_PATH: envalid.str({ default: '/etc/jitsi/shard-status' }),
    POLLING_INTERVAL: envalid.num({ default: 5 }),
    PARTICIPANT_MAX: envalid.num({ default: 5000 }),
});

export default {
    HTTPServerPort: env.HTTP_PORT,
    TCPServerPort: env.TCP_PORT,
    LogLevel: env.LOG_LEVEL,
    JicofoOrig: env.JICOFO_ORIG,
    ProsodyOrig: env.PROSODY_ORIG,
    StatusPath: env.STATUS_PATH,
    PollingInterval: env.POLLING_INTERVAL,
    ParticipantMax: env.PARTICIPANT_MAX,
};
